extensions/pkg/controller/healthcheck/controller.go

```diff
@@ -131,16 +134,6 @@ func (a *AddArgs) GetExtensionGroupVersionKind() schema.GroupVersionKind {
 	return a.registeredExtension.groupVersionKind
 }
 
-// DefaultPredicates returns the default predicates.
-func DefaultPredicates() []predicate.Predicate {
```

Comment:
As the health check is continuously executed (default every 30 seconds) there should be no reason to watch the individual Extension resources. Rather the health check controllers are now only watching the Cluster resource. This is to get updates when the Shoot should be hibernated (stop the health check) and has woken up again (start again).

Reply:
I see.  As we only watch the `Cluster` objects now, how does the cache of the controller-runtime work? Would we now execute an actual `List()` call for the extension objects every time instead of relying on the cache?

Reply:
that would not be good and should be clarified before merging.

Reply:
Added the watch on the extension resource again - so it is the same behaviour as before but an additional watch on the Cluster. Would still be interesting to know if and what the client is caching. It is about[ this code here](https://github.com/gardener/gardener/blob/e408106d170496145fac1abe5cf2e1ccae3088fd/extensions/pkg/controller/healthcheck/reconciler.go#L93). Maybe @tim-ebert can shed some light.

Reply:
Yep, looking into it

Reply:
I think, removing the `DefaultsPredicates` is not a good idea.
Notice the comment, that was there before:
```
// watch: only requeue on spec change to prevent infinite loop	
// health checks are being executed every 'sync period' anyways
```
IIUC and if we remove the default predicates, we will have many unnecessary/unwanted requeues.

Reply:
Wrt to the cache topic, here is what I've found in short:
- under the hood controller-runtime uses `k8s.io/client-go/tools/cache.SharedIndexInformer` for caching
- `controller.Watch(&source.Kind{ ...` will basically result in starting a watch for the given resource via an informer and add an event handler, just like we do for example here:
https://github.com/gardener/gardener/blob/c934fb27dd50489f11330f3fb954fbeb136bd087/pkg/gardenlet/controller/shoot/shoot.go#L119-L126
- `manager.New()` creates two clients
  - one without a cache (direct call to `client.New()`), which can be used for reading directly from the API and injected into controllers/reconcilers by implementing this interface:
https://github.com/gardener/gardener/blob/c934fb27dd50489f11330f3fb954fbeb136bd087/vendor/sigs.k8s.io/controller-runtime/pkg/runtime/inject/inject.go#L45-L47
  - and another one with a cache (if not configured differently via the manager options), which is injected via `InjectClient()` and which we use all over the place in our extensions and extensions library. See for example [here](https://github.com/gardener/gardener-extension-provider-aws/blob/f9b9f455c6af7b6e915f3f2de4a438cb4f663bdf/vendor/sigs.k8s.io/controller-runtime/pkg/manager/manager.go#L317-L332).
This client is a `client.DelegatingClient`, which will use a cache composed of `SharedIndexInformer`s for each `GroupVersionKind`. The trap here is, that for requests with `*unstructured.Unstructured` and `*unstructured.UnstructuredList` it reads directly from the API and not from the cache ([ref](https://github.com/kubernetes-sigs/controller-runtime/blob/fe0e7596fb90a87d024e1d80266b087adbbd36b8/pkg/client/split.go#L46-L61)). All other requests with typed objects/lists, will use the cache, which will be kept in sync by the respective informers.

Reply:
TL;DR:
> Would we now execute an actual List() call for the extension objects every time instead of relying on the cache?

If you mean this call here, no, we would use the client's cache.
https://github.com/gardener/gardener/blob/d311408c6c9a25bc6e204d9599b745168ad2b97b/extensions/pkg/handler/mapper.go#L66

Reply:
The bigger problem is this `Get` call here, as this is an unstructured request and will therefore not use the client's cache:
https://github.com/gardener/gardener/blob/e408106d170496145fac1abe5cf2e1ccae3088fd/extensions/pkg/controller/healthcheck/reconciler.go#L90-L98

Bad news is, this `Get` call is done once per sync period (`30s` as Daniel said) for each extension type, for which a health check is registered.
Good news is, this is easy to fix, as we would only have to replace the first to lines in `Reconcile` by this:
```go
rawExtension := r.registeredExtension.getExtensionObjFunc()
```
Should I open a separate PR for that, as this has nothing to do with this PR?

Reply:
Very nice digging @tim-ebert!
If it's only this one line then it can probably be included in this PR, but it's also fine to keep it separate if @danielfoehrKn prefers that. Please sync with him 👍 

One question, you said that the `SharedIndexInformer`s are created under the hood when creating a new managers, and that such an index informer is used for every `GroupVersionKind`. How does this work internally, i.e., how does the controller-runtime know which objects to cache? According to my understanding, keeping such a cache up-to-date requires a WATCH for the objects, right? I don't believe it starts a WATCH for ALL API objects by default, or? So how does this work?

Reply:
Ok, I'll open a new PR for fixing the get call, as this PR has still open questions, if I see it correctly.